### ISSUE TITLE
All the world

### ISSUE THREAD

**Issue by cdmalcl** [opencog/link-grammar#591]
```
{
numSkippedWords: 2,
linkages: - [
- {
words: - [
"LEFT-WALL",
"a",
"l.u",
"[l]",
"[the]",
"world.n",
"RIGHT-WALL"
],
disjuncts: - [
"Wa+ RW+ ",
"NSa+ ",
"NS- AN+ ",
"",
"",
"@AN- Wa- ",
"RW- "
],
disjunctCost: 0.15000000000000002,
linkageCost: 11,
numViolations: 0,
constituentString: "(S (VP a l.u {l} {the} world.n)) ",
diagramString: " +-----------------RW----------------+ +------------Wa-----------+ | | +NSa+-------AN------+ | | | | | | LEFT-WALL a l.u [l] [the] world.n RIGHT-WALL ",
links: - [
- {
label: "RW",
left: 0,
right: 6,
leftLabel: "RW",
rightLabel: "RW"
},
- {
label: "Wa",
left: 0,
right: 5,
leftLabel: "Wa",
rightLabel: "Wa"
},
- {
label: "AN",
left: 2,
right: 5,
leftLabel: "AN",
rightLabel: "AN"
},
- {
label: "NSa",
left: 1,
right: 2,
leftLabel: "NSa",
rightLabel: "NS"
}
]
},
- {
words: - [
"LEFT-WALL",
"a",
"[l]",
"l.u",
"[the]",
"world.n",
"RIGHT-WALL"
],
disjuncts: - [
"Wa+ RW+ ",
"NSa+ ",
"",
"NS- AN+ ",
"",
"@AN- Wa- ",
"RW- "
],
disjunctCost: 0.15000000000000002,
linkageCost: 11,
numViolations: 0,
constituentString: "(S (VP a {l} l.u {the} world.n)) ",
diagramString: " +----------------RW----------------+ +-----------Wa-----------+ | | +--NSa-+-----AN----+ | | | | | | LEFT-WALL a [l] l.u [the] world.n RIGHT-WALL ",
links: - [
- {
label: "RW",
left: 0,
right: 6,
leftLabel: "RW",
rightLabel: "RW"
},
- {
label: "Wa",
left: 0,
right: 5,
leftLabel: "Wa",
rightLabel: "Wa"
},
- {
label: "AN",
left: 3,
right: 5,
leftLabel: "AN",
rightLabel: "AN"
},
- {
label: "NSa",
left: 1,
right: 3,
leftLabel: "NSa",
rightLabel: "NS"
}
]
}
],
version: "link-grammar-5.3.10",
dictVersion: "5.3.10"
}
```

**Comment by ampli** [comment]
A summary of the posted problem:
Original sentence: `All the world` (this problem doesn't happen with `all the world`).
The produced linkage with LG version 5.3.10 is: `a [l] l.u [the] world.n`

Here is how this happened, with a proposed code fix:

Words which can be separated to words that denote units may cause such a problem.
In our case `All` is separated to `A` `l` ` l` (Amper liter liter).

A similar problem: `As desk or table`
`a [s] desk.n or.j-n table.n`
(Here it got separated as Amper second).

This problem also happens with many other words, like `mall`, `sin` etc.

In the pre-version-5 library, unit separation required a "root number". This requirement has been removed at version 5, which just requires that if the first words is not a unit than it is a number (even though this is not supported by the dictionary - even with a root number multi units are not supported for some reason).

Note that with the latest version (5.4.1) this sentence (`All the world`) doesn't have this problem **by chance**. This is because of a different implementation of null counting got introduced in that version, which prefers the linkage `[all] the world.n`
because it has 1 null word (instead of 2 in `a [l] l.u [the] world.n`).

---

Suggested possible fixes:
1. Require a root number. This at least will prevent most of such bad linkages.
2. On unit separation, issue the separated words with a `.u` subscript. However, I think this can also lead to strange linkages (but I have no example for that). **A problem** with this approach is that not all units in the dictionary have a `.u` subscript , but this can maybe be fixed.

In any case I think this problem should be fixed, and I propose sending a PR implementing option **1** (maybe with **ifdef**).

**Comment by linas** [comment]
Yes, requiring a root number sounds like the right fix.  I have no clue what fix you envision though.  Asking for a root number is like asking for a stem, when splitting off a suffix....

One tricky part is that the "number" might be spelled out:  "four sqft" or "four as" (four ampere-seconds)

**Comment by ampli** [comment]
> One tricky part is that the "number" might be spelled out: "four sqft" or "four as" (four ampere-seconds)

Note that even though `sqft` is currently broken into `sq ft`, the sentence
`We bought 200sq.ft. of plywood.` is doesn't parse when formulated as
`We bought 200sqft of plywood.`
Moreover, I have no evidence that double or more units are supported at all by the current dict.
So I cannot even test that.

**Comment by linas** [comment]
The sqft and related are fixed in #607

**Comment by linas** [comment]
"All the world!" as a single, short response to a question now parses. 

Closing, this will appear as a part of the 5.4.3 release, whenever that comes out.

**Comment by ampli** [comment]
I think that some fix to problems due to the aggressive unit separation is still needed.
Else strange parses happen in many cases. For example: `This is: As`
is parsed as
```
    +--------Xx--------+
    +----->WV----->+   +--Jp--+
    +-->Wd---+-Ss*b+   |  +NSa+
    |        |     |   |  |   |
LEFT-WALL this.p is.v :.j a  s.u
```
Possible solutions:
1. Always require a root digit (already implemented but unpublished).
But then valid constructs like number spelled out in words (including using words like "half") will not be supported. So it is maybe not be a good idea.
2. Issue the units with their subscript, so bogus parses in which the split units are used as non-unit words will not be possible.  To that end, first there is a need to ensure that all the units in the dictionary end with 
`.u` (or possibly with additional known subscripts).
3. Support 2 kinds of links in the dictionary: inter-word and intra-word. The inter-word links will automatically cause morphology splits. In this way the `NSa` above will not be allowed since it is an intar-word link while the words result from separated morphemes. This is not a suggestion for immediate implementation...

**I can implemented (2)** if it looks like a good idea. I checked that by now, all, the relevant units have a `.u` subscript or for some reason (I don't have a guess why) a `.i` subscript. So the unit separation code can issue the separated units with their subscript `.u` or `.i` as they appear in the dict.

**Comment by ampli** [comment]
BTW, the example parse above demonstrate another problem: The separated "A" was converted to "a" while it should have been a final token. This particular bogus parse would not happen if this is fixed.
I added this to the list of problems that need a fix.

**Comment by linas** [comment]
As of just now, all units end with `.u`  However, Having the C code look at the units seems to break the general principles.

In the long run, we need to distinguish inter-word and intra-word links.  Somewhere, we treat LL specially, for Russian.  For English, its murkier:

"you asked for 5 so I gave it to you: 5cu.in."

means that `ND` and `A` are allowed to be "intra-word", but not exclusively. Other examples use `Dmcn`, but never with `.u` units:  "here it is: 5 feet"

**Comment by linas** [comment]
You wrote:
> Moreover, I have no evidence that double or more units are supported at all by the current dict.

They're not, with the exception of sq and cu as abbreviations for square and cubic (which was added in the last 12 hours)

**Comment by ampli** [comment]
> As of just now, all units end with .u However, Having the C code look at the units seems to break the general principles.

We can look at it from another perspective:
The UNITS definition in the affix file is intended for the code which separate units.
The code which reads the affix files strip their subscripts. This is a loss of needed information - I think this may be considered to be against general principles.
I propose to leave them in the UNITS class as they originally are - without stripping.
The unit separation code will then issue them as they are - with subscripts.
This way the parsing ambiguity caused by the loss of subscripts will be solved.]

> In the long run, we need to distinguish inter-word and intra-word links.

Note that this enables getting rid of the regex mechanism (in a part or maybe even totally), provided we invent a way (i.e. a syntax) to represent a character-class.

> For English, its murkier

Generally, we have 3 kinds of links between tokens:
1. Inter-word.
2. Intra-word.
3. Both.

So the link between numbers and units is of the 3 kind.
(BTW, we can also use the '#' way (the fix suggestion - I don't have a better name for it for now...) to denote something like "extra space" if something like e.g. `% 10` is specified.)

**Comment by ampli** [comment]
I said:
>I propose to leave them in the UNITS class as they originally are - without stripping.

If we don't strip subscripts in the affix file, and issue the words in the affix-class as they are, one may wonder how the dummy subscripts in LPUNC etc. will not get issued (without a special handling for affix-classes other than UNITS).

The answer is that these dummy subscripts are not needed at all since when reading the affix file (starting early in version 5) word uniqueness is not enforced. So these subscripts can just be removed (and this will even make the definitions somewhat clearer).

**Comment by linas** [comment]
> these dummy subscripts are not needed .. So these subscripts can just be removed
OK.  Will you remove them? Should I?

> UNITS
OK.

> getting rid of the regex mechanism
I'm confused. I assume you mean, the prefix/suffix regex mechanism? And not the `4.0.regex` file.

> the '#' way
What is this? I guess you've described it in some other bug report, but I don't recall.

**Comment by ampli** [comment]
>>these dummy subscripts are not needed .. So these subscripts can just be removed

>OK. Will you remove them? Should I?

>>UNITS

>OK.

I will do it along with the PR that issues the affix-class words as they are (including subscripts), that I can send soon, to finally solve the aggressive unit stripping issue.

>> getting rid of the regex mechanis

> I'm confused. I assume you mean, the prefix/suffix regex mechanism? And not the 4.0.regex file.

This is according to a proposal I made long ago. If we implement inter-word links, using them we can define what is now implemented using the regex file, and get rid of it.
For example, `NUMBERS`.
(In this example, `[]` is a notation for a character class, and '<'  / '>'s are  '-' / '+' inter-word links. For this example I also supposed an extension in which macros can get a subscript...)
```
[0-9]: <digit>
<digit>.digit: {DN>} & {DN<};
<digit>.!NUMBER: {DN<} & <disjunct-for-number>
% Period and comma can be added too.
```

The display code will see the special subscript prefix `!` and replace the DN-concatenated digits with a "match class"  `NUMBER`, so the result will be exactly like it is now. (Of course DN will be with length_limit=1).

In order that this will work "automagically", we also need to implement a dict-driven tokenization (e.g. by a token tree that will be build by analyzing the inter-links in the dict when the dict is read.)

>> the '#' way

> What is this? I guess you've described it in some other bug report, but I don't recall.

See issue [#404](https://github.com/opencog/link-grammar/issues/404#issuecomment-250018892).

**Comment by linas** [comment]
OK, so the BNF suggestion is interesting:
```
[0-9]: <digit>
<digit>.digit: {DN>} & {DN<};
<digit>.!NUMBER: {DN<} & <disjunct-for-number>
% Period and comma can be added too.
```

I want to say that I like it .. but ... several issues: regexes are compiled, and thus likely to be faster than letting the parser do this.  But also, if I look at 4.0.regex, I see a lot of stuff that would need to be ported, and it becomes a big mess: `ING-WORDS ED-WORDS LY-WORDS` and especially `UNKNOWN-WORD`.

Also `HMS-TIME` might be messy, DAY-ORDINALS too and ROMAN-NUMERAL-WORDS

So I kind-of like this idea but am concerned it becomes big, slow, messy.   Should open a different issue to talk about this, if you wish.

**Comment by ampli** [comment]
>I will do it along with the PR that issues the affix-class words as they are (including subscripts), that I can send soon, to finally solve the aggressive unit stripping issue.

The needed changes were totally minor. But encountered a problem, even though it works fine.

According to the way it was originally written, all the separated tokens but the first one point to the affix class tokens (the first one is recomputed from the original word, because it was supposed to be a root number, a requirement that got omitted later). In this way the first one doesn't get a subscript.

So the sentence: `We bought 200sq.ft. of plywood.` doesn't parse with my modified code. Here is the reason:
```

Debug: Sentence words and alternatives:
...
  word2: bought
   alt0: bought
  word3: 200sq.ft.
   alt0: 200 sq..u ft..u
  word6: of
   alt0: of
...
```
See word3/alt0: we have there `sq..u`. But the dict requires `sq..a`

This sentence parses OK: `We bought 200 sq.ft. of plywood.`
```
...
  word3: 200
   alt0: 200
  word4: sq.ft.
   alt0: sq. {.} ft..u
      alt1: sq
  word7: of
   alt0: of
...
```

This is because now the word is `sq.ft.` and its first token is issued as  `sq.` (a string from the word and not the affix class, as said above).

This demonstrates that the **multi unit** separation in the current code is supposed to work **on a very strange basis**: We are given a list of units from the unit files (as the affix class `UNITS`), essentially all end with the `.u` subscript, but they are supposed sometimes to "match" dictionary tokens ended with other subscripts like `.a`. It seems like an "inappropriate" English knowledge in the code.

This could be solved if we could derive a "token tree" from the dictionary, to be used by the tokenizer.  But currently we have no real basis to know that, for example, `sq..a` should be separated from the string `sq.ft.` (why `sq..u` can be considered as a hint?). To do that (i.e derive a token-tree from the dict), we at least need intra-link marking and/or word boundary marking on the relevant words.

Possibilities:
1) Leave the unit separation as is, knowing that we sometimes get strange parses due to it, until we have a new tokenizer.
2) Use a similar code like my said modified one, with an additional modification to always issue the first unit (of a multi-unit word) without a subscript, on the basis that some units may act as "adjectives" for other units. This has the benefit of reducing some of the ambiguity (by still issuing some of the units with a subscript, especially when there is only one).
3) Issue all units with subscripts, and add the "adjective" ones to the `UNITS` affix class! It seems this solves all the current problems.

I'm propose to choose possibility (3), even though the code change is not a minor one (there will be a need to generate alternatives in the unit separation code if a unit appears with more than one subscript).

**Comment by ampli** [comment]
A specific proposal for implementing (3) above:
Create a new file `/en/words/units.a`:
```
sq.a
sq..a
cu.a
cu..a
```
In `en/4.0.dict.m4`, make this change:
```
-sq.a sq..a cu.a cu..a: A+;
+/en/words/units.a: A+;
```
In `en/4.0.affix` add this new file:
```
+/en/words/units.a: UNITS+;
```
In `tokenize.c`, issue all units with subscripts, using generating alternatives when needed.

**Comment by linas** [comment]
Yes, that sounds fine!

**Comment by ampli** [comment]
Implemented in PR #617.
